File: clients/python/tests/chunking/resources/typescript/sample.ts
/**
 * Sample TypeScript file for testing code chunking.
 *
 * This file contains various TypeScript constructs to test AST parsing.
 */

// Constants
const MAX_RETRIES: number = 3;
const DEFAULT_TIMEOUT: number = 30.0;

// Type definitions
type UserId = string;
type Metadata = Record<string, unknown>;

// Interface definitions
interface IUser {
    id: UserId;
    name: string;
    email?: string;
}

interface IService {
    initialize(): Promise<void>;
    isReady(): boolean;
}

interface IUserService extends IService {
    createUser(id: string, name: string, email?: string): User;
    getUser(id: string): User | null;
    deleteUser(id: string): boolean;
}

// Enum definition
enum ServiceStatus {
    PENDING = 'pending',
    RUNNING = 'running',
    STOPPED = 'stopped',
    ERROR = 'error',
}

/**
 * Represents a user in the system.
 */
class User implements IUser {
    public id: UserId;
    public name: string;
    public email?: string;

    constructor(id: UserId, name: string, email?: string) {
        this.id = id;
        this.name = name;
        this.email = email;
    }

    public getDisplayName(): string {
        return this.name || this.email || this.id;
    }

    public setEmail(email: string): void {
        this.email = email;
    }
}

/**
 * Base class for services.
 */
abstract class BaseService implements IService {
    protected config: Metadata;
    protected _initialized: boolean = false;
    protected status: ServiceStatus = ServiceStatus.PENDING;

    constructor(config: Metadata) {
        this.config = config;
    }

    public async initialize(): Promise<void> {
        this._initialized = true;
        this.status = ServiceStatus.RUNNING;
    }

    public isReady(): boolean {
        return this._initialized;
    }

    protected validateConfig(): boolean {
        return Boolean(this.config);
    }
}

/**
 * Service for managing users.
 */
class UserService extends BaseService implements IUserService {
    private users: Map<string, User> = new Map();

    constructor(config: Metadata) {
        super(config);
    }

    public createUser(id: string, name: string, email?: string): User {
        if (!id) {
            throw new Error('ID cannot be empty');
        }
        const user = new User(id, name, email);
        this.users.set(id, user);
        this.onUserCreated(user);
        return user;
    }

    public getUser(id: string): User | null {
        return this.users.get(id) || null;
    }

    public deleteUser(id: string): boolean {
        return this.users.delete(id);
    }

    private onUserCreated(user: User): void {
        // Internal callback
    }
}

/**
 * Calculate factorial of n.
 */
function calculateFactorial(n: number): number {
    if (n <= 1) {
        return 1;
    }
    return n * calculateFactorial(n - 1);
}

/**
 * Fetch data from URL asynchronously.
 */
async function fetchData(url: string, timeout: number = DEFAULT_TIMEOUT): Promise<Metadata> {
    // Simulated async fetch
    return { url, status: 'ok' };
}

/**
 * Process items with optional validation.
 */
const processItems = (items: string[], validate: boolean = true): string[] => {
    const filtered = validate ? items.filter(item => item) : items;
    return filtered.map(item => item.trim().toLowerCase());
};

// Generic function
function identity<T>(value: T): T {
    return value;
}

// Generic class
class Container<T> {
    private value: T;

    constructor(value: T) {
        this.value = value;
    }

    public getValue(): T {
        return this.value;
    }

    public setValue(value: T): void {
        this.value = value;
    }
}

// Type guard
function isUser(obj: unknown): obj is User {
    return obj instanceof User;
}

// Decorator (experimental)
function logged(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function (...args: any[]) {
        console.log(`Calling ${propertyKey} with`, args);
        return originalMethod.apply(this, args);
    };
    return descriptor;
}

// Main execution
async function main(): Promise<void> {
    const service = new UserService({ env: 'test' });
    await service.initialize();

    const user = service.createUser('1', 'Test User', 'test@example.com');
    console.log(`Created user: ${user.getDisplayName()}`);

    const result = calculateFactorial(5);
    console.log(`Factorial: ${result}`);
}

export {
    User,
    UserService,
    BaseService,
    calculateFactorial,
    fetchData,
    processItems,
    identity,
    Container,
    isUser,
    ServiceStatus,
    MAX_RETRIES,
    DEFAULT_TIMEOUT,
};
export type { IUser, IService, IUserService, UserId, Metadata };

File: clients/python/tests/chunking/resources/ruby/sample.rb
# frozen_string_literal: true

# Sample Ruby file for testing code chunking.
#
# This file contains various Ruby constructs to test AST parsing.

require 'json'
require 'ostruct'

# Constants
MAX_RETRIES = 3
DEFAULT_TIMEOUT = 30.0

# Custom error class
class ServiceError < StandardError
  attr_reader :code

  def initialize(message, code = nil)
    super(message)
    @code = code
  end
end

# Represents a user in the system
class User
  attr_reader :id, :name
  attr_accessor :email

  def initialize(id, name, email = nil)
    @id = id
    @name = name
    @email = email
  end

  def display_name
    name || email || id
  end

  def to_h
    { id: id, name: name, email: email }
  end

  def to_json(*_args)
    to_h.to_json
  end
end

# Base class for services
class BaseService
  attr_reader :config

  def initialize(config = {})
    @config = config
    @initialized = false
  end

  def initialize!
    @initialized = true
  end

  def ready?
    @initialized
  end

  protected

  def validate_config
    !config.empty?
  end
end

# Service for managing users
class UserService < BaseService
  def initialize(config = {})
    super
    @users = {}
  end

  def create_user(id, name, email = nil)
    raise ServiceError.new('ID cannot be empty') if id.nil? || id.empty?

    user = User.new(id, name, email)
    @users[id] = user
    on_user_created(user)
    user
  end

  def get_user(id)
    @users[id]
  end

  def delete_user(id)
    !!@users.delete(id)
  end

  def each_user(&block)
    @users.values.each(&block)
  end

  private

  def on_user_created(user)
    # Internal callback
  end
end

# Module with utility methods
module MathUtils
  module_function

  def calculate_factorial(n)
    return 1 if n <= 1
    n * calculate_factorial(n - 1)
  end

  def fibonacci(n)
    return n if n <= 1
    fibonacci(n - 1) + fibonacci(n - 2)
  end
end

# Mixin module
module Serializable
  def serialize
    to_h.to_json
  end

  def self.included(base)
    base.extend(ClassMethods)
  end

  module ClassMethods
    def from_json(json)
      data = JSON.parse(json, symbolize_names: true)
      new(**data)
    end
  end
end

# Fetch data asynchronously (simulated)
def fetch_data(url, timeout: DEFAULT_TIMEOUT)
  # Simulated fetch
  { url: url, status: 'ok', timeout: timeout }
end

# Process items with optional validation
def process_items(items, validate: true)
  filtered = validate ? items.compact.reject(&:empty?) : items
  filtered.map { |item| item.strip.downcase }
end

# Block and yield example
def with_retry(max_retries = MAX_RETRIES)
  retries = 0
  begin
    yield
  rescue StandardError => e
    retries += 1
    retry if retries < max_retries
    raise e
  end
end

# Struct definition
UserStruct = Struct.new(:id, :name, :email) do
  def display_name
    name || email || id
  end
end

# Main execution
if __FILE__ == $PROGRAM_NAME
  service = UserService.new(env: 'test')
  service.initialize!

  user = service.create_user('1', 'Test User', 'test@example.com')
  puts "Created user: #{user.display_name}"

  result = MathUtils.calculate_factorial(5)
  puts "Factorial: #{result}"
end
